### PR TITLE
`ColorPalette`: refine custom color button's label

### PR DESCRIPTION
**Pull request by ciampo**
## What?

Change the label for the custom color button in the `ColorPalette` component:

- format the color value to always be upper case, and hide the starting `#` character
- display the color name

<details>

<summary>See more details</summary>

For reference, this is how this button looks like in `trunk`

image (1)

And this is how it _should_ look like:

image
</details>

## Why?

Iterating and refining our UIs

## How?

1. Added a utility function that is able to find the color name, given its hex value, by matching it against the color palette(s)
2. Used the `Flex` and `Truncate` components (and a couple of custom styles) to implement the layout changes
3. Added a fix to `fieldset` to make sure that it does't grow larger than its container.

## Testing Instructions

- Check the component in Storybook / Global Styles (in the theme editor) / Post editor (e.g. by setting a background color for a group block)
- Focus your attention on the custom color button, in the `ColorPalette` component:
    - The color value should be uppercase and stripped from the leading `#` character
    - The color name should be displayed. If the current color can't be matched against the current paletted of predefined named colors, the text `Custom` should be displayed instead
    - Make sure that, in case of a color with a very long name, the text is truncated correctly.

## Screenshots or screencast 

**In Storybook:**

https://user-images.githubusercontent.com/1083581/157878761-41d43388-b72b-4ed1-9921-95815e1e3c77.mp4

**In the theme editor:**

https://user-images.githubusercontent.com/1083581/157874005-6f997ffc-43d3-4b0c-894e-12114aa88a6f.mp4

**In the post editor:**

https://user-images.githubusercontent.com/1083581/157875711-94cb9db8-7f80-4f8e-99ce-0b9d2d34cd60.mp4

### CODE REVIEW

File: packages/components/src/color-palette/index.js
Comment (ciampo) [inline packages/components/src/color-palette/index.js:159]: Not sure what to when multiple palettes are passed to the `ColorPalette` component.

For now, i've decided to show the `colorPaletteName` (i.e. `Theme`, `Default`, `Custom`) in between parenthesis after the color name. This was mainly triggered by the fact that `Theme` colors have names like "primary", "secondary", "background"...

Other options could be:

- do not display the `colorPaletteName`
- ignore colors coming from the `Theme` palette when trying to match them with a name (although, are we guaranteed that the theme palette is always going to be called `Theme`? Will it change in other locales?)

File: packages/components/src/color-palette/index.js
Comment (ciampo) [inline packages/components/src/color-palette/index.js]: Not sure what to when multiple palettes are passed to the `ColorPalette` component.

For now, i've decided to show the `colorPaletteName` (i.e. `Theme`, `Default`, `Custom`) in between parenthesis after the color name. This was mainly triggered by the fact that `Theme` colors have names like "primary", "secondary", "background"...

Other options could be:

- do not display the `colorPaletteName`
- ignore colors coming from the `Theme` palette when trying to match them with a name (although, are we guaranteed that the theme palette is always going to be called `Theme`? Will it change in other locales?)

Comment (ciampo) [inline packages/components/src/color-palette/index.js]: Is there a need to mark this text for translation?

File: packages/components/src/color-palette/test/index.js
Comment (ciampo) [inline packages/components/src/color-palette/test/index.js:95]: Unit tests were failing, I believe because the switch from a vanilla `button` to a `Flex as={button}` component meant that Enzyme's shallow render couldn't find that `button` anymore. Changing to `mount` solves the problems (with the consequence of generating a larger snapshot).

Not sure if this is the cleanest solution, but I figured it'd be good enough — especially considering that we should eventually rewrite these tests to use `@testing-library` instead.

**Comment by jasmussen** [comment]
Nice one. This is what I see:
Screenshot 2022-03-11 at 15 30 28

I'll take a look and see if I can't get the padding to match this:

Screenshot 2022-03-11 at 15 31 23

Do we need to include the (Theme)? I feel like there's plenty of context in the sectioning off of the swatches.

**Comment by mtias** [comment]
Yes, let's try without the (Theme)

File: packages/components/src/color-palette/index.js
Comment (mtias) [reply packages/components/src/color-palette/index.js]: It's good to show the name, it helps convey that themes should not be using names like "Background" which are really Elements and not palette items.

If the color that is set is a custom one without a name we should say "Custom"

**Comment by jasmussen** [comment]
For the spacings, here's the source design:
Screenshot 2022-03-11 at 15 43 33

Screenshot 2022-03-11 at 15 44 24

The card is 64px tall, and the red padding as denoted is 12px.

The card inside the flyout is a bit wider, that's okay, the height and padding should still be the same:
Screenshot 2022-03-11 at 15 45 53

**Comment by ciampo** [comment]
Thank you for the quick round of feedback!

I went ahead and implemented it:

1. Tweaked styles (height, padding, vertical alignment) https://github.com/WordPress/gutenberg/pull/39386/commits/48b3822c26d6089df7447ff844df862a0896c8b2
2. Removed color palette name https://github.com/WordPress/gutenberg/pull/39386/commits/4f98c7cdc069707a10329e105f8f86ded96f09a7


<details>

<summary>A few screenshots:</summary>


Screenshot 2022-03-11 at 16 08 52

Screenshot 2022-03-11 at 16 09 40

Screenshot 2022-03-11 at 16 09 58

</details>

**Comment by ciampo** [comment]
Also flagging some failing e2e tests that seem related to the changes in this PR.

I'll still have to wait until next week to merge, since currently e2e tests are broken across all PRs.

**Comment by jasmussen** [comment]
Nice, the metrics are right now!

Screenshot 2022-03-11 at 16 09 04

For that reason, I'm happy to give this a thumbs up. 

The unset state looks odd now, though — it seems like the previous height was informed by the transparency grid size:
Screenshot 2022-03-11 at 16 09 00

Fundamentally I consider that a separate problem, one that is tracked in #36998, and so it probably doesn't need to be fixed by this PR. But it does feel like something to address sooner rather than later.

**Review by jasmussen — APPROVED** [review]

File: packages/components/src/color-palette/index.js
Comment (mirka) [inline packages/components/src/color-palette/index.js]: Did you originally intend to assign a value to this somewhere? `colorNameToReturn` is never set in this code block.

File: packages/components/src/color-palette/index.js
Comment (mirka) [inline packages/components/src/color-palette/index.js]: Non-blocking nit: Because the body of these loops are the same in both conditions, I wonder if we should normalize the arrays earlier to simplify the code? For example:

```js
const colorPalettes = showMultiplePalettes ? colors : [ { colors } ];

for ( const { colors: colorPaletteColors } of colorPalettes ) {
	for ( const {
		name: colorName,
		color: colorValue,
	} of colorPaletteColors ) {
		if (
			colord( currentValue ).toHex() ===
			colord( colorValue ).toHex()
		) {
			return colorName;
		}
	}
}
```

File: packages/components/CHANGELOG.md
Comment (mirka) [inline packages/components/CHANGELOG.md]:
```suggestion
-   `ColorPalette`: refine custom color button's label. ([#39386](https://github.com/WordPress/gutenberg/pull/39386))
```

File: packages/components/src/color-palette/index.js
Comment (mirka) [inline packages/components/src/color-palette/index.js:219]: Kind of similar to [this comment](https://github.com/WordPress/gutenberg/pull/37769#discussion_r815799705) — should the button's aria-label include the current color name and value?

**Review by mirka — COMMENTED** [review]
Looks very nice ✨ 

One tiny thing: In the Storybook, the mock color data has lowercase `name`s. Just as a stylistic thing, could we capitalize those mock color names? (I hope most of the color data in actual usage already have capitalized names. Otherwise we need to consider a text-transform!)

File: packages/base-styles/_variables.scss
Comment (ntsekouras) [inline packages/base-styles/_variables.scss]: Nit: spaces are not consistent.

**Review by ntsekouras — APPROVED** [review]
This looks so good Marco! Thanks! 💯 

I'm approving this but I think @mirka's comments about the unused `colorNameToReturn` and the normalization of input would help to simplify the code.

File: packages/components/src/color-palette/index.js
Comment (ciampo) [reply packages/components/src/color-palette/index.js]: Good point — That was a left-over of a refactor 😅 

Clean up in https://github.com/WordPress/gutenberg/pull/39386/commits/0e585aded465aea7a0bd48ed65583c30d0f1dedf

File: packages/components/src/color-palette/index.js
Comment (ciampo) [reply packages/components/src/color-palette/index.js]: Such a neat idea! Applied in https://github.com/WordPress/gutenberg/pull/39386/commits/73defe321e1dd5c0e54af422a2cf59a7cbb4a9c1

File: packages/components/src/color-palette/index.js
Comment (ciampo) [reply packages/components/src/color-palette/index.js:219]: I gave your idea a try in https://github.com/WordPress/gutenberg/pull/39386/commits/ba6bdd436e0d117bf53a23fd214a7e574bb00e69 and https://github.com/WordPress/gutenberg/pull/39386/commits/bbe98743ef11d53a0dcbb1e52820a2c760f1d53d — although I'm not 10)% sure about the wording that I chose.

Maybe @alexstine can give some feedback on whether the new `aria-label` needs any further improvements?

File: packages/base-styles/_variables.scss
Comment (ciampo) [reply packages/base-styles/_variables.scss]: Well spotted! I swapped spaces with tabs in https://github.com/WordPress/gutenberg/pull/39386/commits/1a89543ba69bd6d7461c3831e980efdee0aefc86

**Comment by ciampo** [comment]
The [`Heading › should correctly apply custom colors` test](https://github.com/WordPress/gutenberg/blob/08759a972d74b214853ebc23267cbc9f6db41c6b/packages/e2e-tests/specs/editor/blocks/heading.test.js#L91) has been failing consistently on this PR, although when I test it locally it keeps passing.

Any ideas about why this could happen?

**Comment by alexstine** [comment]
@ciampo Try to make your environment as close to the real thing as possible.

1. npm ci
2. npm run build
3. npx wp-env start
4. npm run test-e2e -- /path/to/test/

Sometimes I think E2E tests have a mind of their own. Also, you may want to try to run it a couple times. Possible that GitHub Actions server is just running slow. Every now and then, I get green checks from doing nothing. 👍 

I'll accessibility review this later today.

**Comment by ntsekouras** [comment]
>The [Heading › should correctly apply custom colors test](https://github.com/WordPress/gutenberg/blob/08759a972d74b214853ebc23267cbc9f6db41c6b/packages/e2e-tests/specs/editor/blocks/heading.test.js#L91) has been failing consistently on this PR, although when I test it locally it keeps passing.

> Any ideas about why this could happen?

For me it was failing locally as well. Now the `button` has no text nodes, so it seems it has to do with internals of xpath and how handles nested nodes. Here is [something I read](https://stackoverflow.com/questions/3655549/xpath-containstext-some-string-doesnt-work-when-used-with-node-with-more) that helped a bit.

**Comment by ciampo** [comment]
Thank you both! Thanks to @alexstine 's notes, with `npm ci && npm run build` I managed to replicate the failure, which was then addressed by @ntsekouras 's fix (I wrongly assumed that `text()` and `contains()` would be able to read text content from nested nodes 😅 )

File: packages/components/src/color-palette/index.js
Comment (alexstine) [reply packages/components/src/color-palette/index.js:219]: @ciampo Are you able to make the button contain what the default color is without any colors selected?

I am not sure why the color picker opens much lower in the DOM. It used to not do this if memory serves me right but this PR didn't change that. Chocking that up to a regression that happened at some point. This is going to need a considerable amount of work after this gets merged. Something else to focus on.

Comment (ciampo) [reply packages/components/src/color-palette/index.js:219]: > Are you able to make the button contain what the default color is without any colors selected?

I'm not sure I understand what you're exactly referring to. When no color is "picked", I believe that the `value` prop is either `undefined` (or an empty string) — in other words, this component doesn't have a "default" color. At the moment, when the `value` prop is not defined, the aria-label for this `button` should be "Custom color picker."

> I am not sure why the color picker opens much lower in the DOM. It used to not do this if memory serves me right but this PR didn't change that. Chocking that up to a regression that happened at some point. This is going to need a considerable amount of work after this gets merged. Something else to focus on.

I haven't really worked on this component before, but I'm going to ping @jorgefilipecosta and @ntsekouras in case they have a better understanding of the matter that you're highlighting here.

Comment (alexstine) [reply packages/components/src/color-palette/index.js:219]: @ciampo I figured the component had a default value when no color was selected. If not, this PR is good to go on accessibility end.

**Review by alexstine — APPROVED** [review]
Functionality and accessibility seem sane to me. 👍 Thanks @ciampo !

**Comment by ciampo** [comment]
Update:

- Rebased to solve CHANGELOG conflicts
- Also [updated Storybook to controls](https://github.com/WordPress/gutenberg/pull/39386/commits/0c53767ba2fe9179b085c13bf81bd589ab009f03), while including latest piece of feedback from @mirka (about capitalising color names)

E2E tests flakiness seems to be solved, so hopefully I'll be able to merge once all CI checks complete

**Comment by mtias** [comment]
Thank you very much for the refinement and attention to detail!

**Pull request merged**
